### ISSUE TITLE
[naming-convention] add a modifier to identify a variable/parameter as having been destructured

### ISSUE THREAD

**Issue by Timmmm** [typescript-eslint/typescript-eslint#2512]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/naming-convention": "error"
  }
}
```

```TS
import electron from "electron";
const { BrowserWindow } = electron;
```

**Expected Result**

It shouldn't complain about the name of `BrowserWindow`, because I didn't have any choice in the matter.

**Actual Result**

It does complain.

**Additional Info**

I feel like there should be an option to don't check destructured names at all.

**Versions**

All latest.

**Comment by bradzacher** [comment]
> because I didn't have any choice in the matter

You could just as easily write:

```ts
const browserWindow = electron.BrowserWindow;
// or
const { BrowserWindow: browserWindow } = electron;
```

You have a choice of the name, it's just that the simplest syntax reuses the existing name.

---

Destructuring creates variables. These variables should adhere to your naming convention.
Otherwise you will introduce names that do not match your conventions into your codebase.

Side note that there is currently a bug with destructured names: #2250.

**Comment by Timmmm** [comment]
> Destructuring creates variables.

Right, but variables that I didn't name. If I *did* name them (because the come from my codebase) then I'll get a warning at the place where they are defined anyway. [The Rust developers agree](https://github.com/rust-lang/rust/issues/66362#issuecomment-554069636), if that persuades you!

Maybe this should be an option.

**Comment by Timmmm** [comment]
This is actually even worse if you have the `no-unused-vars` rule turned on. If I try to destructure a `[number, number]` array for example:

```
const [ _x, y ] = getCoords();
// _x is unused
```

Even though I have set up `no-unused-vars` to ignore variables starting with `_`, it still complains about it, *and* it complains that the naming convention is wrong. With exactly the same config, this works fine with no errors:

```
const a = getCoords();
const _x = a[0];
const y = a[1];
```

**Comment by Timmmm** [comment]
Actually, scratch that last message - I think it is behaving as intended, it's just that the `_x is assigned a value which is never read` is a bit awkward when used with destructuring.

The naming convention thing was my mistake (forgot to turn on `leadingUnderscore: true` for one selector).

**Comment by bradzacher** [comment]
> The Rust developers agree, if that persuades you!

It doesn't - pattern matching is a bit different. I don't think there's a way to rename the variable there (I don't know anything about rust syntax but I couldn't see it in a quick scan of the docs).

---

Using PascalCase when your naming convention specified camelCase doesn't seem like to egregious of an issue.

But consider that ignoring destructuring naming would allow this:

```ts
const { mY_rId_ICULOUS_ly_$$$_named_variable } = {};
```

It's hard to draw that line and says "this is acceptable variance".
IMO it's better to just say "all variables must follow the naming convention, no matter their source".

Perhaps we could mark these variables with a modifier, and allow you to reconfigure it as strict as you like?

**Comment by Timmmm** [comment]
> I don't think there's a way to rename the variable there

There is - pattern matching is very similar to destructuring in Rust.

> You could instead write
> ```
>     match pet {
>        Animal::Spider { iHateSnakes: i_hate_snakes } => dbg!(i_hate_snakes),
>    };
>```

> Perhaps we could mark these variables with a modifier, and allow you to reconfigure it as strict as you like?

That sounds like a perfect solution!
